source/game/system/map/MapdataCheckPoint.cc

```diff
@@ -24,11 +25,113 @@ void MapdataCheckPoint::read(EGG::Stream &stream) {
     m_nextPt = stream.read_u8();
 }
 
+/// @brief Calculates @ref m_nextPoints and @ref m_prevPoints from @ref m_nextPt and @ref m_prevPt.
+/// @details Also calculates the quadrilaterals for the next checkpoints, filling the fields of @ref
+/// LinkedCheckpoint for each.
+/// @addr{0x80515624}
+void MapdataCheckPoint::initCheckpointLinks(MapdataCheckPointAccessor &accessor, int id) {
+    m_id = id;
+    auto courseMap = CourseMap::Instance();
+
+    // Calculate the quadrilateral's `m_prevPoints`. If the check point is the first in its group
+    // (prev == -1), it has multiple previous checkpoints defined by its preceding checkpaths
+    if (m_prevPt == 0xff) {
+        MapdataCheckPath *checkpath = courseMap->checkPath()->findCheckpathForCheckpoint(id);
+        m_prevCount = 0;
+
+        for (size_t i = 0; i < 6; i++) {
+            u16 prevID = checkpath->getPrev(i);
+
+            if (prevID == 0xff) {
+                continue;
+            }
+
+            MapdataCheckPath *prev = courseMap->checkPath()->get(prevID);
+            m_prevPoints[i] = accessor.get(prev->end());
+            m_prevCount++;
+        }
+    } else {
+        m_prevPoints[0] = accessor.get(m_prevPt);
+        m_prevCount++;
+    }
+
+    // Calculate the quadrilateral's `m_nextPoints`. If the checkpoint is the last in its group, it
+    // can have multiple quadrilaterals (and nextCheckpoint) which are determined by its next
+    // path(s)
+    if (m_nextPt == 0xff) {
+        MapdataCheckPath *checkpath = courseMap->checkPath()->findCheckpathForCheckpoint(id);
+        m_nextCount = 0;
+
+        for (size_t i = 0; i < 6; i++) {
+            u16 nextID = checkpath->getNext(i);
+
+            if (nextID == 0xff) {
+                continue;
+            }
+
+            MapdataCheckPath *next = courseMap->checkPath()->get(nextID);
+            m_nextPoints[i].checkpoint = accessor.get(next->start());
+            m_nextCount++;
+        }
+    } else {
+        m_nextPoints[0].checkpoint = accessor.get(m_nextPt);
+        m_nextCount++;
+    }
+
+    // Form the checkpoint's quadrilateral(s)
+    for (size_t i = 0; i < 6; i++) {
```

Comment:
See above

Reply:
unsure how to convert this one into a range-based foreach loop. The size of `m_nextPoints` is 6, but not all of them are actually valid; the size is instead determined by `m_nextCount`; hence the bounds check. see [this discord discussion](https://discord.com/channels/1230981272139075765/1231411876550803506/1310056608499236894)

Reply:
If I understand correctly, you just said the `6` is derived based on `m_nextCount` being 6. So just put `m_nextCount` then